File: packages/react-native-audio-recorder/ios/WAVFileWriter.swift
import Foundation

class WAVFileWriter {

  private let filePath: String
  private let sampleRate: Double
  private let channels: Int

  private var fileHandle: FileHandle?
  private var totalSamplesWritten: Int = 0

  init(filePath: String, sampleRate: Double, channels: Int) throws {
    self.filePath = filePath
    self.sampleRate = sampleRate
    self.channels = channels

    // Create file
    FileManager.default.createFile(atPath: filePath, contents: nil)

    // Open file handle
    fileHandle = try FileHandle(forWritingTo: URL(fileURLWithPath: filePath))

    // Write WAV header (placeholder, will update on finalize)
    try writeWAVHeader(dataSize: 0)
  }

  func write(samples: UnsafeMutablePointer<Int16>, count: Int) {
    guard let fileHandle = fileHandle else { return }

    let data = Data(bytes: samples, count: count * MemoryLayout<Int16>.size)
    fileHandle.write(data)

    totalSamplesWritten += count
  }

  func finalize() throws -> RecordingResult {
    guard let fileHandle = fileHandle else {
      throw NSError(domain: "WAVFileWriter", code: -1, userInfo: [
        NSLocalizedDescriptionKey: "File handle is nil"
      ])
    }

    // Update header with correct data size
    try fileHandle.seek(toOffset: 0)
    let dataSize = totalSamplesWritten * MemoryLayout<Int16>.size
    try writeWAVHeader(dataSize: dataSize)

    // Close file
    try fileHandle.close()
    self.fileHandle = nil

    // Get file size
    let attributes = try FileManager.default.attributesOfItem(atPath: filePath)
    let fileSize = attributes[.size] as! Int

    // Calculate duration
    let durationSeconds = Double(totalSamplesWritten) / (sampleRate * Double(channels))
    let durationMs = Int(durationSeconds * 1000)

    return RecordingResult(
      filePath: filePath,
      durationMs: durationMs,
      fileSizeBytes: fileSize,
      sampleRate: sampleRate,
      channels: channels
    )
  }

  func cancel() {
    try? fileHandle?.close()
    try? FileManager.default.removeItem(atPath: filePath)
  }

  private func writeWAVHeader(dataSize: Int) throws {
    var header = Data()

    // RIFF chunk
    header.append("RIFF".data(using: .ascii)!)
    header.append(UInt32(36 + dataSize).littleEndian.data)
    header.append("WAVE".data(using: .ascii)!)

    // fmt chunk
    header.append("fmt ".data(using: .ascii)!)
    header.append(UInt32(16).littleEndian.data) // Chunk size
    header.append(UInt16(1).littleEndian.data)  // Audio format (PCM)
    header.append(UInt16(channels).littleEndian.data)
    header.append(UInt32(sampleRate).littleEndian.data)

    let byteRate = Int(sampleRate) * channels * 2 // 16-bit = 2 bytes
    header.append(UInt32(byteRate).littleEndian.data)

    let blockAlign = channels * 2
    header.append(UInt16(blockAlign).littleEndian.data)
    header.append(UInt16(16).littleEndian.data) // Bits per sample

    // data chunk
    header.append("data".data(using: .ascii)!)
    header.append(UInt32(dataSize).littleEndian.data)

    fileHandle?.write(header)
  }
}

struct RecordingResult {
  let filePath: String
  let durationMs: Int
  let fileSizeBytes: Int
  let sampleRate: Double
  let channels: Int
}

extension FixedWidthInteger {
  var data: Data {
    var value = self
    return Data(bytes: &value, count: MemoryLayout<Self>.size)
  }
}

File: packages/react-native-audio-recorder/android/src/main/java/com/choiminseok/audiorecorder/WAVFileWriter.kt
package com.choiminseok.audiorecorder

import java.io.File
import java.io.FileOutputStream
import java.io.RandomAccessFile
import java.nio.ByteBuffer
import java.nio.ByteOrder

class WAVFileWriter(
    private val filePath: String,
    private val sampleRate: Int,
    private val channels: Int
) {
    private var fileOutputStream: FileOutputStream? = null
    private var totalSamplesWritten: Int = 0
    private val file: File = File(filePath)

    init {
        // Create file
        file.createNewFile()

        // Open file output stream
        fileOutputStream = FileOutputStream(file)

        // Write WAV header (placeholder, will update on finalize)
        writeWAVHeader(0)
    }

    fun write(samples: ShortArray, count: Int) {
        val outputStream = fileOutputStream ?: return

        // Convert shorts to bytes
        val byteBuffer = ByteBuffer.allocate(count * 2)
        byteBuffer.order(ByteOrder.LITTLE_ENDIAN)

        for (i in 0 until count) {
            byteBuffer.putShort(samples[i])
        }

        outputStream.write(byteBuffer.array())
        totalSamplesWritten += count
    }

    fun finalize(): RecordingResult {
        val outputStream = fileOutputStream ?: throw IllegalStateException("File output stream is null")

        // Close output stream
        outputStream.close()
        fileOutputStream = null

        // Update header with correct data size
        val dataSize = totalSamplesWritten * 2 // 2 bytes per sample (16-bit)
        updateWAVHeader(dataSize)

        // Get file size
        val fileSize = file.length().toInt()

        // Calculate duration
        val durationSeconds = totalSamplesWritten.toDouble() / (sampleRate * channels)
        val durationMs = (durationSeconds * 1000).toInt()

        return RecordingResult(
            filePath = filePath,
            durationMs = durationMs,
            fileSizeBytes = fileSize,
            sampleRate = sampleRate,
            channels = channels
        )
    }

    fun cancel() {
        fileOutputStream?.close()
        fileOutputStream = null
        file.delete()
    }

    private fun writeWAVHeader(dataSize: Int) {
        val outputStream = fileOutputStream ?: return

        val header = ByteBuffer.allocate(44)
        header.order(ByteOrder.LITTLE_ENDIAN)

        // RIFF chunk
        header.put("RIFF".toByteArray())
        header.putInt(36 + dataSize)
        header.put("WAVE".toByteArray())

        // fmt chunk
        header.put("fmt ".toByteArray())
        header.putInt(16) // Chunk size
        header.putShort(1) // Audio format (PCM)
        header.putShort(channels.toShort())
        header.putInt(sampleRate)

        val byteRate = sampleRate * channels * 2 // 16-bit = 2 bytes
        header.putInt(byteRate)

        val blockAlign = (channels * 2).toShort()
        header.putShort(blockAlign)
        header.putShort(16) // Bits per sample

        // data chunk
        header.put("data".toByteArray())
        header.putInt(dataSize)

        outputStream.write(header.array())
    }

    private fun updateWAVHeader(dataSize: Int) {
        RandomAccessFile(file, "rw").use { raf ->
            // Update file size (RIFF chunk size)
            raf.seek(4)
            raf.write(intToByteArray(36 + dataSize))

            // Update data chunk size
            raf.seek(40)
            raf.write(intToByteArray(dataSize))
        }
    }

    private fun intToByteArray(value: Int): ByteArray {
        return byteArrayOf(
            (value and 0xFF).toByte(),
            ((value shr 8) and 0xFF).toByte(),
            ((value shr 16) and 0xFF).toByte(),
            ((value shr 24) and 0xFF).toByte()
        )
    }
}
